server/authentication_test.go

```diff
@@ -0,0 +1,134 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Marc Berhault (marc@cockroachlabs.com)
+
+package server
+
+import (
+	"fmt"
+	"net/http"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/kv"
+	"github.com/cockroachdb/cockroach/testutils"
+	"github.com/cockroachdb/cockroach/ts"
+)
+
+// Verify client certificate enforcement.
+func TestSSLEnforcement(t *testing.T) {
+	s := StartTestServer(t)
+	defer s.Stop()
+	testCases := []struct {
+		method, key   string
+		certsStatus   int // Status code for https with client certs.
+		noCertsStatus int // Status code for https without client certs.
+	}{
+		// /ui/: basic file server: no auth.
+		{"GET", "/index.html", http.StatusOK, http.StatusOK},
+
+		// /_admin/: server.adminServer: no auth.
+		{"GET", healthPath, http.StatusOK, http.StatusOK},
+
+		// /debug/: server.adminServer: no auth.
+		{"GET", debugEndpoint + "vars", http.StatusOK, http.StatusOK},
+
+		// /_status/: server.statusServer: no auth.
+		{"GET", statusKeyPrefix, http.StatusOK, http.StatusOK},
+
+		// /kv/rest/: kv.RESTServer. Key cannot be empty. Response is NotFound since the key does not exist.
+		{"GET", kv.EntryPrefix + "foo", http.StatusNotFound, http.StatusUnauthorized},
+
+		// /kv/db/: kv.DBServer. These are proto reqs, but we can at least tell get past auth.
+		{"GET", kv.DBPrefix + "Get", http.StatusBadRequest, http.StatusUnauthorized},
+
+		// /ts/: ts.Server.
+		{"GET", ts.URLPrefix, http.StatusNotFound, http.StatusUnauthorized},
+	}
+
+	// HTTPS with client certs.
+	certsContext := testutils.NewTestBaseContext()
+	client, err := certsContext.GetHTTPClient()
+	if err != nil {
+		t.Fatalf("error initializing http client: %s", err)
+	}
+	for tcNum, tc := range testCases {
+		req, err := http.NewRequest(tc.method,
+			fmt.Sprintf("%s://%s%s", certsContext.RequestScheme(), s.ServingAddr(), tc.key),
+			nil)
+		if err != nil {
+			t.Fatalf("[%d]: error building request: %s", tcNum, err)
+		}
+
+		resp, err := client.Do(req)
+		if err != nil {
+			t.Fatalf("[%d]: %s: error issuing request: %s", tcNum, req.URL, err)
+		}
+
+		defer resp.Body.Close()
+		if resp.StatusCode != tc.certsStatus {
+			t.Errorf("[%d]: %s: expected status code %d, got %d", tcNum, req.URL, tc.certsStatus, resp.StatusCode)
+		}
+	}
+
+	// HTTPS without client certs.
+	noCertsContext := testutils.NewTestBaseContext()
+	noCertsContext.Certs = ""
+	client, err = noCertsContext.GetHTTPClient()
+	if err != nil {
+		t.Fatalf("error initializing http client: %s", err)
+	}
+	for tcNum, tc := range testCases {
```

Comment:
nit: do this in one iteration above (i.e. make each request twice before moving to the next one) or extract this into a helper